pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/FSTBasedRegexpPredicateEvaluatorFactory.java

```diff
@@ -0,0 +1,124 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.filter.predicate;
+
+
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import org.apache.pinot.core.query.request.context.predicate.Predicate;
+import org.apache.pinot.core.segment.index.readers.Dictionary;
+import org.apache.pinot.core.segment.index.readers.TextIndexReader;
+import org.apache.pinot.core.util.fst.RegexpMatcher;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+
+import static org.apache.pinot.core.query.request.context.predicate.Predicate.Type.REGEXP_LIKE;
+
+public class FSTBasedRegexpPredicateEvaluatorFactory {
+    public FSTBasedRegexpPredicateEvaluatorFactory() {}
+
+    public static BaseDictionaryBasedPredicateEvaluator newFSTBasedEvaluator(
+            TextIndexReader fstIndexReader, Dictionary dictionary, String regexpQuery) {
+        return new FSTBasedRegexpPredicateEvaluatorFactory.FSTBasedRegexpPredicateEvaluator(
+                fstIndexReader, dictionary, regexpQuery);
+    }
+
+    public static BaseDictionaryBasedPredicateEvaluator newAutomatonBasedEvaluator(
+            Dictionary dictionary, String regexpQuery) {
+        return new FSTBasedRegexpPredicateEvaluatorFactory.AutomatonBasedRegexpPredicateEvaluator(
+                regexpQuery, dictionary);
+    }
+
+    private static class AutomatonBasedRegexpPredicateEvaluator extends BaseDictionaryBasedPredicateEvaluator {
+        private final RegexpMatcher _regexpMatcher;
+        private final Dictionary _dictionary;
+        int[] _matchingDictIds;
+
+        public AutomatonBasedRegexpPredicateEvaluator(String searchQuery, Dictionary dictionary) {
+            _regexpMatcher = new RegexpMatcher(searchQuery, null);
+            _dictionary = dictionary;
+        }
+
+        @Override
+        public Predicate.Type getPredicateType() {
+            return REGEXP_LIKE;
+        }
+
+        @Override
+        public boolean applySV(int dictId) {
+            return _regexpMatcher.match(_dictionary.getStringValue(dictId));
+        }
+
+        @Override
+        public int[] getMatchingDictIds() {
+            if (_matchingDictIds == null) {
+                IntList matchingDictIds = new IntArrayList();
+                int dictionarySize = _dictionary.length();
+                for (int dictId = 0; dictId < dictionarySize; dictId++) {
+                    if (applySV(dictId)) {
+                        matchingDictIds.add(dictId);
+                    }
+                }
+                _matchingDictIds = matchingDictIds.toIntArray();
+            }
+            return _matchingDictIds;
+        }
+    }
+
+    private static class FSTBasedRegexpPredicateEvaluator extends BaseDictionaryBasedPredicateEvaluator {
```

Comment:
What's the difference between both predicate evaluators?

Reply:
FSTBasedRegexpPredicateEvaluator is used for rolled out segments and AutomatonBasedRegexpPredicateEvaluator for consuming segments. Reason for this is because java regexp matcher is much broader than the regexp matcher I am borrowing from lucene. So, to keep the results consistent between consuming and rolled out segments I am using similar matching logic when fst index is enabled. 

To reduce complexity here we could either:
* switch regexp_like to use this matching using lucene lib everywhere 
or
* create a new operator.

Maybe in future if there is a native automaton building library, this will simplify things.

Reply:
Added comments too, let me know if I should expand more